File: OLD/javascript/selection-sort.js
const { generateShuffledArray, generateExpectedArray } =  require("./arrays")

const _selectionSort = (arr) =>
{
    const src = arr.slice(0)
    const res = []
    for (let i = 0; i < arr.length; i++) {
        let min = src[0]
        let pos = 0
        let j = 0
        while (j < src.length) {
            if (src[j] < min) {
                min = src[j]
                pos = j
            }
            j++
        }
        res.push(min) // Add min to result array
        src.splice(pos, 1) // Remove min from src
    }
    return res
}

const selectionSort = (arr) =>
{
    const start = new Date().getTime()
    const resultArray = _selectionSort(arr)
    const end = new Date().getTime()
    return {
        res: resultArray,
        sortTime: end - start
    }
}

const main = () =>
{
    const size = 40000

    // Generate array
    const { arr, genTime } = generateShuffledArray(size)

    // Sort shuffled array
    const { res, sortTime } = selectionSort(arr)

    // Eval result array with an expected result
    const expected = generateExpectedArray(size)
    const isSortedAsExpected = res.toString() == expected.toString()

    // Output the results
    console.log(`Time to generate shuffled array of size ${size}: ${genTime} ms`)
    if (! isSortedAsExpected) {
        console.log("2. The result arrays is NOT sorted as expected")
    } else {
        console.log("2. The result arrays is sorted as expected")
        console.log(`3. Time to selection sort the shuffled array: ${sortTime} ms`)
    }
}

main()

File: OLD/cpp/src/selection-sort.cpp
#include "../headers/Arrays.h"
#include "../headers/Time.h"

#include <cmath>

using std::cout;

std::shared_ptr<int[]> _selection_sort(std::shared_ptr<int[]> src, const int size)
{
    std::shared_ptr<int[]> res(new int[size]);
    for (int i = 0; i < size; i++) {
        int min_pos = 0;
        int j = 0;
        const int src_size = size - i;
        while (j < src_size) {
            if (src[j] < src[min_pos]) min_pos = j;
            j++;
        }
        res[i] = src[min_pos];
        for (int k = min_pos; k < src_size; k++) {
            src[k] = src[k + 1];
        }
    }
    return res;
}

Result selection_sort(const std::shared_ptr<int[]> arr, const int size)
{
    auto arr_copy = copy_array(arr, size);
    auto start = clock();
    auto res = _selection_sort(arr_copy, size);
    auto end = clock();
    auto sortTime = get_time_ms(start, end);
    return (Result) {res, sortTime};
}

int main()
{
    const int size = 40000;

    // Generate shuffled array
    auto genResult = generate_shuffled_array(size);

    // Sort shuffled array
    auto sortResult = selection_sort(genResult.arr, size);

    // Eval sorted array with an expected result
    auto expected = generate_expected_array(size);

    // Output the results
    cout << "1. Time to generate array of size " << size << ": "
         << floor(genResult.time) << " ms\n";
    if (! is_sorted_as_expected(sortResult.arr, expected, size)) {
        cout << "2. The result array is NOT sorted as expected\n";
    } else {
        cout << "2. The result array is sorted as expected\n";
        cout << "3. Time to selection sort the array: "
             << floor(sortResult.time) << " ms\n";
    }

    return 0;
}
